feat: add cache abstraction for fetching signing keys

sreya opened the pull request:
This PR adds the database implementation for fetching and caching keys used for JWT signing. It's been merged into the `keyrotate` pkg and renamed to `cryptokeys` since they're coupled concepts. Given the size of the PR, an additional PR will be added for the wsproxy implementation.

sreya committed:
feat: add keychain abstraction for fetching rotated keys

sreya committed:
Refactor DBKeyCache to DBCache with clock support

- Simplify naming by renaming `DBKeyCache` to `DBCache`.
- Introduce `DBCacheOption` for flexible clock configuration.
- Ensure consistent clock handling across the methods.

sreya committed:
Refactor CryptoKey handling to use codersdk package

Refactored the CryptoKey handling logic to leverage the codersdk
package. This enhances consistency and modularity by centralizing
CryptoKey management and related operations. The refactoring also
includes renaming functions to better reflect their functionality
and using the db2sdk package for database to SDK conversions.

sreya committed:
Refactor CryptoKey conversion to use db2sdk package

- Simplifies conversion and improves consistency by using
  a common package for translating database keys to SDK keys.
- Improves test reliability and code maintainability by centralizing
  the conversion process.

sreya committed:
make gen

sreya committed:
Refactor key verification with db2sdk conversion

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,185 @@
+
+func (d *DBCache) refresh(ctx context.Context) {
+	d.clock.TickerFunc(ctx, time.Minute*10, func() error {
+		cache, latest, err := d.newCache(ctx)
```
I can't help but wonder whether it might be better / simpler to use an `AfterFunc` timer to invalidate the cache with a 10 minute timeout.  If we do a fetch at 9m55s, it doesn't make a lot of sense to do another fetch 5s later. (Whenever we fetch, reset the timer to 10m.)

Also, we can just leave the cache empty on startup and on invalidation: the next call to grab a key will populate it.

sreya replied:
Agreed

sreya replied:
I think this matters less for the DB implementation but I think for the WSProxy one (which I'll be putting up once this gets merged) we would want to avoid only invalidating the cache since it's guarantees you'll always have a slow request

spikecurtis reviewed (commented):
You don't have to convert all the times to UTC. Go handles timezones just fine without this.

sreya commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,185 @@
+	d.cacheMu.Lock()
+	defer d.cacheMu.Unlock()
+
+	if latest.CanSign(now) {
```
🤦 so bad, good catch

sreya commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,185 @@
+		return checkKey(key, now)
+	}
+
+	key, err := d.db.GetCryptoKeyByFeatureAndSequence(ctx, database.GetCryptoKeyByFeatureAndSequenceParams{
```
In a similar vein we're not really saving much by remembering the `latestkey`, wondering if we should just compute on each call

spikecurtis replied:
Up to you.  You're right that spinning thru a map of like 2 keys will probably take less than a microsecond.

sreya committed:
pr comments

sreya committed:
Refactor error message for key fetching

sreya commented on coderd/database/db2sdk/db2sdk.go:
```diff
@@ -659,3 +659,17 @@ func Organization(organization database.Organization) codersdk.Organization {
+	return codersdk.CryptoKey{
+		Feature:   codersdk.CryptoKeyFeature(key.Feature),
+		Sequence:  key.Sequence,
+		StartsAt:  key.StartsAt.UTC(),
```
The main motivatino here is `testify` which is being rather obnoxious about trying to do `Equal` comparisons between objects because there's a TZ discrepancy. E.g. 

```bash
        	Error:      	[]codersdk.CryptoKey{codersdk.CryptoKey{Feature:"workspace_apps", Secret:"462543a101f54c2aa51cbeeede315ef2d5a1bb6b15eb7f3df6c0a948e6ce64999650d7effc3da276c84d0e5ae34122f4225c2f12df75cacd8752b06faebc414b6ea98e1dba072988836b47fc8f711788c26feed32ed3a4a578c3467b8bfeea7c", DeletesAt:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Sequence:3, StartsAt:time.Date(2024, time.September, 29, 23, 55, 41, 913601000, time.Location(""))}, codersdk.CryptoKey{Feature:"workspace_apps", Secret:"a2f6d56687565f9407ab5710baabb0db82732567e0abbee17846cfd95868a3f0625e7e3ace10e29933d463c3d4068b806205c132a49fdac14fd336db20c0230157ac60bfb8e8067f18e41bddda1f7498374342a9088a33139aa9ac35409d5e8c", DeletesAt:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Sequence:2, StartsAt:time.Date(2024, time.September, 29, 22, 55, 41, 913601000, time.Location(""))}} does not contain codersdk.CryptoKey{Feature:"workspace_apps", Secret:"a2f6d56687565f9407ab5710baabb0db82732567e0abbee17846cfd95868a3f0625e7e3ace10e29933d463c3d4068b806205c132a49fdac14fd336db20c0230157ac60bfb8e8067f18e41bddda1f7498374342a9088a33139aa9ac35409d5e8c", DeletesAt:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Sequence:2, StartsAt:time.Date(2024, time.September, 29, 22, 55, 41, 913601000, time.Location("Canada/Newfoundland"))}
```

spikecurtis replied:
Fine in test code, I guess, but I don't like seeing all the unnecessary `UTC()` calls in the product code.

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,186 @@
+			defer close(fired)
+
+			// There's a small window where the timer fires as we're fetching
+			// keys that could result in us immediately invalidating the cache that we just populated.
```
I'm ok living with this in the product, since it's an extremely narrow window and worst case we just fetch twice instead of once.

However, I'm *not* OK with test flakes due to this issue, so we'll have to consider carefully whether there are any.  It might be safest just to fix it so that we don't leave a lil landmine for future unit test writers.

If you want to fix it, then (given the design I describe below where we reset the timer in `fetch`) what you do is store an `invalidateAt` time on the `DBCache` struct, protected by the mutex. You set it to 10 minutes in the future just _before_ resetting the timer on a `fetch`. In the `AfterFunc` callback, check whether we still want to invalidate by comparing the current time with `invalidateAt` (while holding the mutex).  If a new `fetch` raced with invalidation and won, we'd have moved the time forward and can make the spurious `AfterFunc` callback a no-op.

(You can also have the invalidation call close a channel, and if `reset()` returns false, wait for the channel to close and re-lock, but that seems more complicated to me.)

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,186 @@
+		})
+
+		select {
+		case <-ctx.Done():
```
In order to not leak a goroutine in the timer, you need to stop it when the context completes (or add an explicit `Close()` method to the cache).

Should probably add `goleak` to the test suite.

sreya committed:
Refactor DBKeyCache with robust invalidation timing

Introduce a timer-based cache invalidation system in DBCache to enhance reliability. The new implementation minimizes cache invalidation race conditions during key fetching, ensuring consistent cache state management. Additions include a 'Close' method for releasing resources, such as timers, and test improvements for timer behavior validation.

sreya committed:
Add leak detection to dbkeycache tests

sreya committed:
Add mutex lock to DBCache Close method

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,189 @@
+
+	now := d.clock.Now()
+	d.timer.Stop()
+	d.timer = d.newTimer()
```
instead of stopping the old and creating the new, you can do

```
d.timer.Reset(10*time.Minutes)
```

sreya replied:
I wasn't aware you could just reset it if it had already fired TIL

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,189 @@
+func (d *DBCache) Close() {
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+	d.timer.Stop()
```
So, it's good that you hold the mutex here, but if we need to be concerned with other things calling the `DBCache` at the same time (we do), then we also need to ensure another call doesn't come along after this one, trigger a fetch, and restart the timer.

You could add a `closed` field, or just nil out the timer.  The question is, when to check?  We could check on every call to Signing / Verifying --- basically refuse to answer any more key queries after we are closed.  

That seems a little unfriendly if things are getting closed down in multiple goroutines. It might be nice to serve keys we have locally, but refuse to query the database. The danger there is if we close the DBCache, but something else still has a reference to it, it could happily serve keys forever without erroring.

So, probably the safest thing is to stop answering key queries once we are closed.

sreya committed:
Add closed state to DBCache for safe shutdown

Introduce `closed` state to the DBCache to ensure operations such
as Verifying and Signing are safely terminated when the cache is
closed. This update prevents resource access and ensures proper
shutdown sequences, enhancing system reliability.

sreya committed:
Optimize timer reset in fetch method

sreya committed:
Refactor DBCache to remove unused timer function

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,202 @@
+
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+
```
need to re-check `d.closed`

spikecurtis commented on coderd/cryptokeys/dbkeycache.go:
```diff
@@ -0,0 +1,202 @@
+
+	d.keysMu.Lock()
+	defer d.keysMu.Unlock()
+
```
need to re-check `d.closed`

spikecurtis reviewed (approved):
I don't need to review again, but you need to fix the re-checking of `d.close` when you release the read lock and acquire the write lock.

sreya committed:
finale
